### PR TITLE
Add SPIR-V 1.4 checks

### PR DESCRIPTION
**Pull request by MrSidims**
Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

**Comment by MrSidims** [comment]
- [ ] Update tests

File: lib/SPIRV/libSPIRV/SPIRVValue.cpp [hunk 1]
```diff
@@ -102,11 +101,10 @@ void SPIRVValue::setNoUnsignedWrap(bool HasNoUnsignedWrap) {
     return;
   }
   if (Module->isAllowedToUseExtension(
-          ExtensionID::SPV_KHR_no_integer_wrap_decoration)) {
+          ExtensionID::SPV_KHR_no_integer_wrap_decoration) &&
+      Module->isAllowedToUseVersion(VersionNumber::SPIRV_1_4)) {
```
[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVValue.cpp]: Done

File: lib/SPIRV/libSPIRV/SPIRVValue.cpp [hunk 2]
```diff
@@ -80,11 +80,10 @@ void SPIRVValue::setNoSignedWrap(bool HasNoSignedWrap) {
     eraseDecorate(DecorationNoSignedWrap);
   }
   if (Module->isAllowedToUseExtension(
-          ExtensionID::SPV_KHR_no_integer_wrap_decoration)) {
+          ExtensionID::SPV_KHR_no_integer_wrap_decoration) &&
+      Module->isAllowedToUseVersion(VersionNumber::SPIRV_1_4)) {
```
[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVValue.cpp]: done

**Comment by MrSidims** [comment]
@Fznamznon @mlychkov @svenvh please take a look

File: lib/SPIRV/SPIRVWriter.cpp [hunk 3]
```diff
@@ -3870,7 +3874,9 @@ bool LLVMToSPIRVBase::transExecutionMode() {
       case spv::ExecutionModeSignedZeroInfNanPreserve:
       case spv::ExecutionModeRoundingModeRTE:
       case spv::ExecutionModeRoundingModeRTZ: {
-        if (!BM->isAllowedToUseExtension(ExtensionID::SPV_KHR_float_controls))
+        if (!(BM->isAllowedToUseExtension(
+                  ExtensionID::SPV_KHR_float_controls) &&
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVWriter.cpp:3878]: ||

File: lib/SPIRV/libSPIRV/SPIRVValue.cpp [hunk 4]
```diff
@@ -78,13 +78,18 @@ bool SPIRVValue::hasNoSignedWrap() const {
 void SPIRVValue::setNoSignedWrap(bool HasNoSignedWrap) {
```
[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVValue.cpp]: Done

File: lib/SPIRV/libSPIRV/SPIRVValue.h [hunk 5]
```diff
@@ -98,8 +98,31 @@ class SPIRVValue : public SPIRVEntry {
...
+      return;
+    }
+    // NoSignedWrap and NoUnsignedWrap decorations are available only if it is
+    // allowed to use SPIR-V 1.4 or if SPV_KHR_no_integer_wrap_decoration
+    // extension is enabled
+    if (Module->isAllowedToUseVersion(VersionNumber::SPIRV_1_4)) {
+      Module->setMinSPIRVVersion(
+          static_cast<SPIRVWord>(VersionNumber::SPIRV_1_4));
+      addDecorate(new SPIRVDecorate(NoIntegerWrapDecoration, this));
+      SPIRVDBG(spvdbgs() << "Set nsw for obj " << Id << "\n")
```
Comment (mlychkov) [inline lib/SPIRV/libSPIRV/SPIRVValue.h:115]:
```suggestion
      SPIRVDBG(spvdbgs() << "Set "
                         << (NoIntegerWrapDecoration == DecorationNoSignedWrap) ? "nsw" : "nuw"
                         <<" for obj " << Id << "\n")
```

File: lib/SPIRV/libSPIRV/SPIRVValue.h [hunk 6]
```diff
@@ -98,8 +98,35 @@ class SPIRVValue : public SPIRVEntry {
...
+                   ? "nsw"
+                   : "nuw"
+                         << " for obj " << Id << "\n")
+    } else if (Module->isAllowedToUseExtension(
+                   ExtensionID::SPV_KHR_no_integer_wrap_decoration)) {
+      Module->addExtension(ExtensionID::SPV_KHR_no_integer_wrap_decoration);
+      addDecorate(new SPIRVDecorate(NoIntegerWrapDecoration, this));
+      SPIRVDBG(spvdbgs() << "Set nsw for obj " << Id << "\n")
+    } else {
+      SPIRVDBG(spvdbgs() << "Skip setting nsw for obj " << Id << "\n")
```
Comment (mlychkov) [inline lib/SPIRV/libSPIRV/SPIRVValue.h:126]: Here as well. Maybe better to make a variable for that value?

Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVValue.h]: Applied

**Comment by MrSidims** [comment]
There is an issue reported, which I can't reproduce locally. So I will debug through CI a bit, sorry for spam.

File: lib/SPIRV/libSPIRV/SPIRVValue.h [hunk 7]
```diff
@@ -98,8 +98,34 @@ class SPIRVValue : public SPIRVEntry {
...
+  template <spv::Decoration NoIntegerWrapDecoration>
+  void setNoIntegerDecorationWrap(bool HasNoIntegerWrap) {
+    if (!HasNoIntegerWrap) {
+      eraseDecorate(NoIntegerWrapDecoration);
+      return;
+    }
+    // NoSignedWrap and NoUnsignedWrap decorations are available only if it is
+    // allowed to use SPIR-V 1.4 or if SPV_KHR_no_integer_wrap_decoration
+    // extension is enabled
+    const std::string InstStr =
```
Comment (mlychkov) [inline lib/SPIRV/libSPIRV/SPIRVValue.h]: The last moment here is that maybe it would be better to define InstStr only when SPIRVDBG is defined.

Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVValue.h:112]: Applied, thanks!

**Comment by MrSidims** [comment]
I messed up with the git history a bit, changes from the last patch set: added macro and added spirv-max-version option=1.1 to two of the tests. Will create a translator issue in a moment.

**Comment by MrSidims** [comment]
Created the following issue: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1205

File: lib/SPIRV/SPIRVWriter.cpp [hunk 8]
```diff
@@ -3829,53 +3839,47 @@ bool LLVMToSPIRVBase::transExecutionMode() {
         }
       } break;
       case spv::ExecutionModeNoGlobalOffsetINTEL: {
-        if (BM->isAllowedToUseExtension(
-                ExtensionID::SPV_INTEL_kernel_attributes)) {
-          BF->addExecutionMode(BM->add(
-              new SPIRVExecutionMode(BF, static_cast<ExecutionMode>(EMode))));
-          BM->addCapability(CapabilityKernelAttributesINTEL);
-        }
+        if (!BM->isAllowedToUseExtension(
```
Comment (mlychkov) [inline lib/SPIRV/SPIRVWriter.cpp:3842]: Does this change functional or just a change of coding style?

File: lib/SPIRV/libSPIRV/SPIRVValue.h [hunk 9]
```diff
@@ -98,8 +98,36 @@ class SPIRVValue : public SPIRVEntry {
 
   void setAlignment(SPIRVWord);
   void setVolatile(bool IsVolatile);
-  void setNoSignedWrap(bool HasNoSignedWrap);
-  void setNoUnsignedWrap(bool HasNoUnsignedWrap);
+
+  template <spv::Decoration NoIntegerWrapDecoration>
```
Comment (mlychkov) [inline lib/SPIRV/libSPIRV/SPIRVValue.h:102]: Just a nit: template function definition may be left in cpp file. One can just write explicit function specializations after the definition to eliminate linkage errors.

File: lib/SPIRV/SPIRVWriter.cpp (hunk 8 above)
Comment (MrSidims) [reply lib/SPIRV/SPIRVWriter.cpp:3842]: Just a coding style

File: lib/SPIRV/libSPIRV/SPIRVValue.h (hunk 9 above)
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVValue.h:102]: Oh, I was trying to do that, but failed, but I just realized, that I tried it within a class, but should do specialization outside of the class definition.

File: lib/SPIRV/libSPIRV/SPIRVValue.cpp [hunk 10]
```diff
@@ -124,6 +89,40 @@ void SPIRVValue::setFPFastMathMode(SPIRVWord M) {
                      << "\n")
 }
 
+template <spv::Decoration NoIntegerWrapDecoration>
+void SPIRVValue::setNoIntegerDecorationWrap(bool HasNoIntegerWrap) {
```
Comment (Fznamznon) [inline lib/SPIRV/libSPIRV/SPIRVValue.cpp:93]: What is the purpose of HasNoIntegerWrap parameter? Calls to this functions are added only with true parameter and under if checking that the LLVM value has "NoIntegerWrap".

File: test/transcoding/global_block.cl [hunk 11]
```diff
@@ -4,8 +4,18 @@
...
-// RUN: llvm-spirv %t.bc -o %t.spv
+// TODO: currently max version is limited to 1.1 for this test. Issues here
+// that the SPIR-V module generated for blocks is invalid for versions starting
+// from 1.4, spirv-val is failing with:
+//   error: line 63: Interface variable id <13> is used by entry point
+//                   'block_kernel' id <24>, but is not listed as an interface
+//   %__block_literal_global = OpVariable %_ptr_CrossWorkgroup__struct_10
+//                         CrossWorkgroup %11
+// details can be found in:
+// – Public issue #35: OpEntryPoint must list all global variables in the
```
Comment (Fznamznon) [inline test/transcoding/global_block.cl:15]: Is it a bug in the translator?

**Review by mlychkov — APPROVED** [review]
LGTM, thank you.

File: lib/SPIRV/libSPIRV/SPIRVValue.cpp (hunk 10 above)
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVValue.cpp:93]: I also find this strange, but prefer not to change it within this patch set (even if it already brings some code outside of the declared purpose).

File: test/transcoding/global_block.cl (hunk 11 above)
Comment (MrSidims) [reply test/transcoding/global_block.cl:15]: Yes, see: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1205

File: lib/SPIRV/libSPIRV/SPIRVValue.cpp (hunk 10 above)
Comment (Fznamznon) [reply lib/SPIRV/libSPIRV/SPIRVValue.cpp:93]: :(

**Pull request merged**